siva/metadata.go

```diff
@@ -0,0 +1,278 @@
+	if v != m.CurrentVersion {
+		m.dirty = true
+		m.CurrentVersion = v
+	}
+}
+
+// Save writes metadata to the library yaml file.
+func (m *LibraryMetadata) Save(fs billy.Filesystem) error {
+	data, err := m.yaml()
+	if err != nil {
+		return err
+	}
+
+	tmp := LibraryMetadataFile + ".tmp"
+	defer fs.Remove(tmp)
+
+	err = util.WriteFile(fs, tmp, data, 0666)
+	if err != nil {
+		return err
+	}
+
+	return fs.Rename(tmp, LibraryMetadataFile)
+}
+
+// Dirty returns true if the metadata was changed and it needs to be written.
+func (m *LibraryMetadata) Dirty() bool {
+	return m.dirty
+}
+
+func (m *LibraryMetadata) yaml() ([]byte, error) {
+	return yaml.Marshal(*m)
+}
+
+// parseLibraryMetadata parses the yaml representation of library metadata.
+func parseLibraryMetadata(d []byte) (*LibraryMetadata, error) {
+	var m LibraryMetadata
+
+	err := yaml.Unmarshal(d, &m)
+	if err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+// loadLibraryMetadata reads and parses a library metadata file.
+func loadLibraryMetadata(fs billy.Filesystem) (*LibraryMetadata, error) {
+	mf, err := fs.Open(LibraryMetadataFile)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer mf.Close()
+
+	data, err := ioutil.ReadAll(mf)
+	if err != nil {
+		return nil, err
+	}
+
+	return parseLibraryMetadata(data)
+}
+
+// Version describes a bookmark in the siva file.
+type Version struct {
+	// Offset is a position in the siva file.
+	Offset uint64 `json:"offset"`
+	// Size is block size of the Version.
+	Size uint64 `json:"size,omiempty"`
+}
+
+// LocationMetadata holds extra data associated with a siva file.
+type LocationMetadata struct {
+	// Versions holds a numbered list of bookmarks in the siva file.
+	Versions map[int]Version `json:"versions"`
+
+	dirty bool
+}
+
+// NewLocationMetadata creates a new LocationMetadata.
+func NewLocationMetadata(versions map[int]Version) *LocationMetadata {
+	return &LocationMetadata{
+		Versions: versions,
+	}
+}
+
+// parseLocationMetadata parses the yaml representation of location metadata.
+func parseLocationMetadata(d []byte) (*LocationMetadata, error) {
+	var m LocationMetadata
+
+	err := yaml.Unmarshal(d, &m)
+	if err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+// loadLocationMetadata reads and parses a location metadata file.
+func loadLocationMetadata(
+	fs billy.Filesystem,
+	path string,
+) (*LocationMetadata, error) {
+	mf, err := fs.Open(path)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer mf.Close()
+
+	data, err := ioutil.ReadAll(mf)
+	if err != nil {
+		return nil, err
+	}
+
+	return parseLocationMetadata(data)
+}
+
+// locationMetadataPath returns the path for a location metadata file.
+func locationMetadataPath(path string) string {
+	return path + ".yaml"
+}
+
+// yaml returns the yaml representation of the location metadata.
+func (m *LocationMetadata) yaml() ([]byte, error) {
```

Comment:
I think there's no need to wrap `yaml.Marshal` in a method instead of use it directly.